### ISSUE TITLE
How to get Pester to see custom Should operators?

### ISSUE THREAD

**Issue by kamranayub** [pester/Pester#590]
I must be missing something simple because I feel really close.

I have defined a new `PesterBeDate` function and failure functions. According to what I see in the Should assertion source, it should be able to find my command. In fact if I run:

```
Get-Command PesterBeDate -CommandType Function
```

I see the command output. But for some reason Pester doesn't:

```
RuntimeException: 'BeDate' is not a valid Should operator.
   at Get-TestResult, C:\Program Files\WindowsPowerShell\Modules\Pester\3.4.2\Functions\Assertions\Should.ps1: line 42
```

**Comment by dlwyatt** [comment]
Just put your function into a PS1 file in the Pester\Functions\Assertions directory, and reload the module.  Should be all set.  :)

**Comment by kamranayub** [comment]
Ah, but this for an OSS project... Is there a way to source Pester not from
global?

**Comment by dlwyatt** [comment]
It probably depends on where you've defined those functions.  Unless they're in the global scope, Pester's calls to Get-Command aren't going to find them.  (The way that `Should` operators get loaded is also changing in v4.)

However, for now, you can inject your functions into Pester's scope after loading the module and it should work:

```posh
$pesterModule = Import-Module Pester -PassThru

& $pesterModule {
    function PesterBeDate { }
    # etc
}
```

**Comment by kamranayub** [comment]
Thanks, I did end up sourcing a local module for better out of the box
tests for contributors. Then just copying any assertions after the fact and
that works fine.

**Comment by kamranayub** [comment]
Here's an example of what I did for future reference: https://github.com/kamranayub/posh-openweathermap/blob/afe931f979dd675c279eecd89028f3fe7c4a5893/Run-Tests.ps1

**Comment by ryangribble** [comment]
> (The way that Should operators get loaded is also changing in v4.)

Hey guys, is there any information about how these have changed in v4?  Ive been unsuccesful in located instructions on how to do custom Should extensions in the new version?

**Comment by dlwyatt** [comment]
Didn't write instructions yet, will get to that this weekend.  However, you can look at https://github.com/pester/Pester/blob/master/Functions/Assertions/Be.ps1 for guidance.  Two main changes:

- You have to call `Add-AssertionOperator` to register your operator with Pester.
- Your test function now has to return an object with `Succeeded` (boolean) and `FailureMessage` (string) properties.  `FailureMessage` is ignored unless `Succeeded` is `$false`.
- Your test function, at a minimum, must take three parameters:  `$ActualValue, $ExpectedValue, [switch] $Negate`.  `-Negate` is the equivalent of what we used to do with separate functions (`PesterBe` and `PesterNotBe`, etc.)  If you add any other named parameters to your test function, they'll show up as dynamic parameters when `Should` is called with your operator.

The names of your functions don't matter anymore.  You can add aliases for your operator with the `-Alias` parameter on `Add-AssertionOperator`.  If your operator can handle entire arrays, you can add the `-SupportsArrayInput` switch, and anything that gets piped to `Should` will be passed in as an array to a single call of your operator in the `$ActualValue` parameter.  If you don't use `-SupportsArrayInput`, then piped in elements to `Should` will get tested one at a time, same as how everything worked in v3.

**Comment by dlwyatt** [comment]
Actually, Be.ps1 is probably one of the more complex files.  Here's a simple one:  https://github.com/pester/Pester/blob/master/Functions/Assertions/BeGreaterThan.ps1

**Comment by ryangribble** [comment]
Thanks for the info @dlwyatt 

It doesn't seem like `Add-AssetionOperator` is an exported command though?  Running version 4.0.3

```PS C:\Users\ryan.gribble> Get-Command -Module Pester -Name Get-*

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Function        Get-MockDynamicParameters                          4.0.3      pester
Function        Get-TestDriveItem                                  4.0.3      pester
```

**Comment by dlwyatt** [comment]
That's correct, though there's a PR open right now to change that.

**Comment by stuartleeks** [comment]
I've just installed 4.0.4 to try to update a couple of libraries with custom assertions and am hitting this issue. I can't see `Add-AssertionOperator`

**Comment by nohwnd** [comment]
#765 merged need to release new version

**Comment by stuartleeks** [comment]
Thanks @nohwnd. Do you have any idea of timelines for a new release now that this is merged?

**Comment by it-praktyk** [comment]
@stuartleeks , please read #793.

**Comment by stuartleeks** [comment]
Thanks @it-praktyk, I read through #793 when I was trying to track down my issue. I guess the questions I was really asking is whether there will be a release ahead of a final v4 release (as there still seems to be a list of issues before then) :-)

**Comment by nohwnd** [comment]
@stuartleeks it's released as 4.0.5.

**Comment by stuartleeks** [comment]
Awesome, thanks @nohwnd !
